Clean invalid UTF-8 characters

rubenarakelyan opened the pull request:
This commit adds a class that removes invalid UTF-8 characters from strings. The class is used when parsing keywords passed to finder-frontend in order to fix `ArgumentError: invalid byte sequence in UTF-8` exceptions.

Trello: https://trello.com/c/piGeShKO/388-fix-some-errors-from-finder-frontend-errbit

tijmenb commented on app/lib/utf8_cleaner.rb:
```diff
@@ -0,0 +1,26 @@
+        undef: :replace,
+        replace: ''
+      ).presence
+    else
```
Is govuk-lint running on this project? Couple of things it should've caught.

rubenarakelyan replied:
https://github.com/alphagov/finder-frontend/pull/269 :)

rubenarakelyan replied:
Lint is now in and I've fixed up what it caught

MatMoore commented:
Can you write a couple of tests that use multibyte unicode characters?

How does rails set the encoding of it's parameter strings? It seems like if the string is not currently binary you would strip out extra characters here:

```
irb(main):001:0> "£".encoding
=> #<Encoding:UTF-8>
irb(main):002:0> "£".encode("UTF-8", "binary")
Encoding::UndefinedConversionError: "\xC2" from ASCII-8BIT to UTF-8
```

rubenarakelyan commented:
I've added a test with an emoji character to test handling of multi-byte UTF-8 characters.

I also changed the encode method to go from UTF-8 to UTF-8 (i.e. no change of encoding), since it's the stripping out of invalid characters rather than the change of encoding that makes the real difference here.

rubenarakelyan commented:
I've also added the UTF8Cleaner in a couple of other places where keywords are taken from the query string to prevent more "invalid UTF-8" errors.

MatMoore commented:
No idea what it's used for, but there is also this rack middleware on alphagov, which rejects requests that aren't encoded in UTF-8: https://github.com/alphagov/invalid_utf8_rejector

rubenarakelyan commented:
Ah, I did not know about that gem! I'll close this PR and open a new one that just installs it for finder-frontend. The difference is that this PR strips out invalid characters but keeps the rest, whereas the gem rejects the entire request if it contains any such characters. I don't think this should matter since users won't normally be accessing pages like this - these requests are coming from automated bots.
